cache/demux.go

```diff
@@ -152,18 +152,33 @@ func (d *demux) Broadcast(events []*clientv3.Event) {
 }
 
 // Purge is called when etcd compaction invalidates our cached history, so clients should resubscribe.
-func (d *demux) Purge() {
+func (d *demux) Purge() { d.purge(false, 0) }
+
+// PurgeWithCompaction sends CompactRevision cancels to watchers with too-old next revisions, then stops others to avoid a resync race.
+func (d *demux) PurgeWithCompaction(compactRev int64) { d.purge(true, compactRev) }
+
+func (d *demux) purge(isCompaction bool, compactRev int64) {
```

Comment:
Don't think I like the idea of reusing `purge` function for purging all watchers and compacting watchers. Can you just implement it as separate functions?